xdsl/interpreters/shaped_array.py

```diff
@@ -17,34 +21,64 @@ class ShapedArray(Generic[_T]):
     data: list[_T]
     shape: list[int]
 
+    def __init__(self, data: list[_T] | _T, shape: list[int]):
+        if not isinstance(data, list):
+            data = [data] * prod(shape)
+
+        self.data = data
+        self.shape = shape
+
     def __post__init__(self):
         assert prod(self.shape) == len(self.data)
 
-    def offset(self, indices: tuple[int, ...]) -> int:
+    def offset(self, index: Index) -> int:
         """
         Returns the index of the element in `self.data` for a given tuple of indices
         """
-        if len(indices) != len(self.shape):
-            raise ValueError(f"Invalid indices {indices} for shape {self.shape}")
+        if len(index) != len(self.shape):
+            raise ValueError(f"Invalid indices {index} for shape {self.shape}")
         # For each dimension, the number of elements in the nested arrays
         prod_dims: list[int] = list(
             accumulate(reversed(self.shape), operator.mul, initial=1)
         )[:-1]
-        offsets = map(operator.mul, reversed(indices), prod_dims)
+        offsets = map(operator.mul, reversed(index), prod_dims)
         offset = sum(offsets)
         return offset
 
-    def load(self, indices: tuple[int, ...]) -> _T:
+    def load(self, index: tuple[int, ...]) -> _T:
         """
         Returns the element for a given tuple of indices
         """
-        return self.data[self.offset(indices)]
+        return self.data[self.offset(index)]
 
-    def store(self, indices: tuple[int, ...], value: _T) -> None:
+    def store(self, index: tuple[int, ...], value: _T) -> None:
         """
         Returns the element for a given tuple of indices
         """
-        self.data[self.offset(indices)] = value
+        self.data[self.offset(index)] = value
+
+    def indices(self) -> Iterable[tuple[int, ...]]:
+        """
+        Iterates over the indices of this shaped array.
+        """
+        for indices in product(*(range(dim) for dim in self.shape)):
+            yield tuple(indices)
```

Comment:
`for/yield` can be folded to a `yield from`:

```suggestion
        yield from product(*(range(dim) for dim in self.shape))
```

Reply:
Good spot I thouhgt it wasn't a tuple for some reason